BE crash in TestStreamLoadJson.test_stream_load_json_with_error_json_root

<!-- (At least include the following, feel free to add if you have more content) -->

### Steps to reproduce the behavior (Required)
```
----------------------current_case_index:1 / 1 of case_1.txt------------------
test_stream_load_json.TestStreamLoadJson.test_stream_load_json_with_error_json_root ... create database stream_load_test_db_0463df20_b1cb_11ec_91a8_00163e0e550b
use stream_load_test_db_0463df20_b1cb_11ec_91a8_00163e0e550b
CREATE TABLE `test_nonsense_columns` (
                    `category` varchar(100) NOT NULL COMMENT "",
                    `author` varchar(100) NOT NULL COMMENT "",
                    `title` varchar(100) NOT NULL DEFAULT "0" COMMENT "",
                    `price` double NOT NULL DEFAULT "0" COMMENT ""
                  ) ENGINE=OLAP
                  UNIQUE KEY(`category`)
                  COMMENT "OLAP"
                  DISTRIBUTED BY HASH(`category`) BUCKETS 10
                  PROPERTIES (
                  "replication_num" = "1"
                  );
curl --location-trusted -u root: -T /home/disk5/fuwei/code/work/StarRocksTest/lib/../common/data/stream_load/data_for_json_root -XPUT -H label:stream_load_1648824605189 -H format:json -H columns:category, author, title, price -H json_root:["$.RECORDS"] -H strip_outer_array:true http://172.26.92.139:8072/api/stream_load_test_db_0463df20_b1cb_11ec_91a8_00163e0e550b/test_nonsense_columns/_stream_load
drop database stream_load_test_db_0463df20_b1cb_11ec_91a8_00163e0e550b
FAIL

======================================================================
FAIL: test_stream_load_json.TestStreamLoadJson.test_stream_load_json_with_error_json_root
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/disk5/fuwei/code/work/StarRocksTest/case/system_case/test_dml/test_stream_load/test_stream_load_json.py", line 921, in test_stream_load_json_with_error_json_root
    eq_(res["Status"], "Fail", res["Message"])
AssertionError:   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0   252    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0

100   252    0     0  100   252      0   4221 --:--:-- --:--:-- --:--:--  4221
curl: (52) Empty reply from server


----------------------------------------------------------------------
XML: /home/disk5/fuwei/code/work/StarRocksTest/result/TestStreamLoadJson_test_stream_load_json_with_error_json_root.xml
----------------------------------------------------------------------
Ran 1 test in 0.086s

FAILED (failures=1)
1
--------------------------------------------------------------------------------------------------------------
0
```

### Expected behavior (Required)
Test Success.
### Real behavior (Required)
BE crashes.
```
*** Aborted at 1648824605 (unix time) try "date -d @1648824605" if you are using GNU date ***
PC: @          0x250e013 starrocks::vectorized::ExpandedJsonArrayParserWithRoot::parse()
*** SIGSEGV (@0x0) received by PID 197258 (TID 0x7fe5dec9e700) from PID 0; stack trace: ***
    @          0x3678dd2 google::(anonymous namespace)::FailureSignalHandler()
    @     0x7fe7158e0400 (unknown)
    @          0x250e013 starrocks::vectorized::ExpandedJsonArrayParserWithRoot::parse()
    @          0x2511dab starrocks::vectorized::JsonReader::_read_and_parse_json()
    @          0x2514115 starrocks::vectorized::JsonReader::read_chunk()
    @          0x2514374 starrocks::vectorized::JsonScanner::get_next()
    @          0x2500ca7 starrocks::vectorized::FileScanNode::_scanner_scan()
    @          0x2501e4f starrocks::vectorized::FileScanNode::_scanner_worker()
    @          0x50ab300 execute_native_thread_routine
    @     0x7fe715695ea5 start_thread
    @     0x7fe7159a8b0d __clone
    @                0x0 (unknown)
### StarRocks version (Required)
branch-2.1

--- a/be/src/exec/vectorized/json_parser.cpp
+++ b/be/src/exec/vectorized/json_parser.cpp
@@ -126,7 +126,7 @@ Status JsonDocumentStreamParserWithRoot::get_current(simdjson::ondemand::object*
 
     // json root filter.
     simdjson::ondemand::value val;
-    RETURN_IF_ERROR(JsonFunctions::extract_from_object(*row, _root_paths, val));
+    RETURN_IF_ERROR(JsonFunctions::extract_from_object(*row, _root_paths, &val));
 
     try {
         if (val.type() != simdjson::ondemand::json_type::object) {
@@ -149,7 +149,7 @@ Status JsonArrayParserWithRoot::get_current(simdjson::ondemand::object* row) noe
     RETURN_IF_ERROR(this->JsonArrayParser::get_current(row));
     simdjson::ondemand::value val;
     // json root filter.
-    RETURN_IF_ERROR(JsonFunctions::extract_from_object(*row, _root_paths, val));
+    RETURN_IF_ERROR(JsonFunctions::extract_from_object(*row, _root_paths, &val));
 
     try {
         if (val.type() != simdjson::ondemand::json_type::object) {
@@ -172,7 +172,7 @@ Status ExpandedJsonDocumentStreamParserWithRoot::parse(uint8_t* data, size_t len
     RETURN_IF_ERROR(this->JsonDocumentStreamParser::get_current(&_curr_row));
 
     simdjson::ondemand::value val;
-    RETURN_IF_ERROR(JsonFunctions::extract_from_object(_curr_row, _root_paths, val));
+    RETURN_IF_ERROR(JsonFunctions::extract_from_object(_curr_row, _root_paths, &val));
 
     try {
         if (val.type() != simdjson::ondemand::json_type::array) {
@@ -229,7 +229,7 @@ Status ExpandedJsonDocumentStreamParserWithRoot::advance() noexcept {
             RETURN_IF_ERROR(this->JsonDocumentStreamParser::get_current(&_curr_row));
 
             simdjson::ondemand::value val;
-            RETURN_IF_ERROR(JsonFunctions::extract_from_object(_curr_row, _root_paths, val));
+            RETURN_IF_ERROR(JsonFunctions::extract_from_object(_curr_row, _root_paths, &val));
 
             try {
                 if (val.type() != simdjson::ondemand::json_type::array) {
@@ -259,7 +259,7 @@ Status ExpandedJsonArrayParserWithRoot::parse(uint8_t* data, size_t len, size_t 
     RETURN_IF_ERROR(this->JsonArrayParser::get_current(&_curr_row));
 
     simdjson::ondemand::value val;
-    RETURN_IF_ERROR(JsonFunctions::extract_from_object(_curr_row, _root_paths, val));
+    RETURN_IF_ERROR(JsonFunctions::extract_from_object(_curr_row, _root_paths, &val));
 
     try {
         if (val.type() != simdjson::ondemand::json_type::array) {
@@ -316,7 +316,7 @@ Status ExpandedJsonArrayParserWithRoot::advance() noexcept {
             RETURN_IF_ERROR(this->JsonArrayParser::get_current(&_curr_row));
 
             simdjson::ondemand::value val;
-            RETURN_IF_ERROR(JsonFunctions::extract_from_object(_curr_row, _root_paths, val));
+            RETURN_IF_ERROR(JsonFunctions::extract_from_object(_curr_row, _root_paths, &val));
 
             try {
                 if (val.type() != simdjson::ondemand::json_type::array) {

--- a/be/src/exec/vectorized/json_scanner.cpp
+++ b/be/src/exec/vectorized/json_scanner.cpp
@@ -517,7 +517,7 @@ Status JsonReader::_construct_row(simdjson::ondemand::object* row, Chunk* chunk,
                 row->reset();
                 RETURN_IF_ERROR(add_nullable_column(column, slot_descs[i]->type(), slot_descs[i]->col_name(), row,
                                                     !_strict_mode));
-            } else if (!JsonFunctions::extract_from_object(*row, _scanner->_json_paths[i], val).ok()) {
+            } else if (!JsonFunctions::extract_from_object(*row, _scanner->_json_paths[i], &val).ok()) {
                 if (strcmp(column_name, "__op") == 0) {
                     // special treatment for __op column, fill default value '0' rather than null
                     if (column->is_binary()) {

--- a/be/src/exprs/vectorized/json_functions.cpp
+++ b/be/src/exprs/vectorized/json_functions.cpp
@@ -110,7 +110,7 @@ Status JsonFunctions::json_path_close(starrocks_udf::FunctionContext* context,
 }
 
 Status JsonFunctions::extract_from_object(simdjson::ondemand::object& obj, const std::vector<SimpleJsonPath>& jsonpath,
-                                          simdjson::ondemand::value& value) noexcept {
+                                          simdjson::ondemand::value* value) noexcept {
     if (jsonpath.size() <= 1) {
         // The first elem of json path should be '$'.
         // A valid json path's size is >= 2.
@@ -174,7 +174,7 @@ Status JsonFunctions::extract_from_object(simdjson::ondemand::object& obj, const
         }
     }
 
-    std::swap(value, tvalue);
+    std::swap(*value, tvalue);
 
     return Status::OK();
 }
@@ -257,7 +257,7 @@ ColumnPtr JsonFunctions::_iterate_rows(FunctionContext* context, const Columns& 
             }
 
             simdjson::ondemand::value value;
-            auto st = extract_from_object(obj, jsonpath, value);
+            auto st = extract_from_object(obj, jsonpath, &value);
             if (!st.ok()) {
                 result.append_null();
                 continue;
@@ -297,7 +297,7 @@ ColumnPtr JsonFunctions::_iterate_rows(FunctionContext* context, const Columns& 
                 }
 
                 simdjson::ondemand::value value;
-                auto st = extract_from_object(obj, jsonpath, value);
+                auto st = extract_from_object(obj, jsonpath, &value);
                 if (!st.ok()) {
                     result.append_null();
                     continue;

--- a/be/src/exprs/vectorized/json_functions.cpp
+++ b/be/src/exprs/vectorized/json_functions.cpp
@@ -111,6 +111,12 @@ Status JsonFunctions::json_path_close(starrocks_udf::FunctionContext* context,
 
 Status JsonFunctions::extract_from_object(simdjson::ondemand::object& obj, const std::vector<SimpleJsonPath>& jsonpath,
                                           simdjson::ondemand::value& value) noexcept {
+    if (jsonpath.size() <= 1) {
+        // The first elem of json path should be '$'.
+        // A valid json path's size is >= 2.
+        return Status::InvalidArgument("empty json path");
+    }
+
     simdjson::ondemand::value tvalue;
 
     // Skip the first $.
@@ -122,6 +128,8 @@ Status JsonFunctions::extract_from_object(simdjson::ondemand::object& obj, const
         const std::string& col = jsonpath[i].key;
         int index = jsonpath[i].idx;
 
+        // Since the simdjson::ondemand::object cannot be converted to simdjson::ondemand::value,
+        // we have to do some special treatment for the second elem of json path.
         if (i == 1) {
             auto err = obj.find_field_unordered(col).get(tvalue);
             if (err) {
@@ -137,23 +145,32 @@ Status JsonFunctions::extract_from_object(simdjson::ondemand::object& obj, const
         }
 
         if (index != -1) {
-            auto arr = tvalue.get_array();
-            if (arr.error()) {
+            // try to access tvalue as array.
+            simdjson::ondemand::array arr;
+            auto err = tvalue.get_array().get(arr);
+            if (err) {
                 return Status::InvalidArgument(fmt::format("failed to access field as array, field: {}, err: {}", col,
-                                                           simdjson::error_message(arr.error())));
+                                                           simdjson::error_message(err)));
             }
 
-            int idx = 0;
-            for (auto a : arr) {
-                if (a.error()) {
-                    return Status::InvalidArgument(fmt::format("failed to access array element, field: {}, err: {}",
-                                                               col, simdjson::error_message(arr.error())));
-                }
+            size_t sz;
+            err = arr.count_elements().get(sz);
+            if (err) {
+                return Status::InvalidArgument(
+                        fmt::format("failed to get array size, field: {}, err: {}", col, simdjson::error_message(err)));
+            }
 
-                if (idx++ == index) {
-                    a.get(tvalue);
-                    break;
-                }
+            if (index >= sz) {
+                return Status::InvalidArgument(
+                        fmt::format("the acquired index is beyond array size, field: {}, index: {}, array size: {}",
+                                    col, index, sz));
+            }
+
+            err = arr.at(index).get(tvalue);
+            if (err) {
+                return Status::InvalidArgument(
+                        fmt::format("failed to access array, field: {}, index: {}, array size: {}, err: {}", col, index,
+                                    sz, simdjson::error_message(err)));
             }
         }
     }

--- a/be/src/exprs/vectorized/json_functions.cpp
+++ b/be/src/exprs/vectorized/json_functions.cpp
@@ -161,9 +161,8 @@ Status JsonFunctions::extract_from_object(simdjson::ondemand::object& obj, const
             }
 
             if (index >= sz) {
-                return Status::InvalidArgument(
-                        fmt::format("the acquired index is beyond array size, field: {}, index: {}, array size: {}",
-                                    col, index, sz));
+                return Status::NotFound(
+                        fmt::format("index beyond array size, field: {}, index: {}, array size: {}", col, index, sz));
             }
 
             err = arr.at(index).get(tvalue);

--- a/be/src/exprs/vectorized/json_functions.h
+++ b/be/src/exprs/vectorized/json_functions.h
@@ -172,7 +172,7 @@ public:
     // extract_from_object extracts value from object according to the json path.
     // Now, we do not support complete functions of json path.
     static Status extract_from_object(simdjson::ondemand::object& obj, const std::vector<SimpleJsonPath>& jsonpath,
-                                      simdjson::ondemand::value& value) noexcept;
+                                      simdjson::ondemand::value* value) noexcept;
 
     static void parse_json_paths(const std::string& path_strings, std::vector<SimpleJsonPath>* parsed_paths);
 

--- a/be/test/exprs/vectorized/json_functions_test.cpp
+++ b/be/test/exprs/vectorized/json_functions_test.cpp
@@ -646,7 +646,7 @@ TEST_F(JsonFunctionsTest, extract_from_object_test) {
     EXPECT_STREQ(output.data(), "2");
 
     st = test_extract_from_object(R"({"data": [{"key": 1},{"key": 2}]})", "$.data[2].key", &output);
-    EXPECT_STATUS(Status::InvalidArgument(""), st);
+    EXPECT_STATUS(Status::NotFound(""), st);
 }
 
 } // namespace vectorized

--- a/be/test/exprs/vectorized/json_functions_test.cpp
+++ b/be/test/exprs/vectorized/json_functions_test.cpp
@@ -32,6 +32,29 @@ public:
         expr_node.type = gen_type_desc(TPrimitiveType::BOOLEAN);
     }
 
+    Status test_extract_from_object(std::string input, std::string jsonpath, std::string* output) {
+        // reverse for padding.
+        input.reserve(input.size() + simdjson::SIMDJSON_PADDING);
+
+        simdjson::ondemand::parser parser;
+        simdjson::ondemand::document doc;
+        EXPECT_EQ(simdjson::error_code::SUCCESS, parser.iterate(input).get(doc));
+
+        simdjson::ondemand::object obj;
+
+        EXPECT_EQ(simdjson::error_code::SUCCESS, doc.get_object().get(obj));
+
+        std::vector<SimpleJsonPath> path;
+        JsonFunctions::parse_json_paths(jsonpath, &path);
+
+        simdjson::ondemand::value val;
+        RETURN_IF_ERROR(JsonFunctions::extract_from_object(obj, path, val));
+        std::string_view sv = simdjson::to_json_string(val);
+
+        output->assign(sv.data(), sv.size());
+        return Status::OK();
+    }
+
 public:
     TExprNode expr_node;
 };
@@ -606,6 +629,25 @@ INSTANTIATE_TEST_SUITE_P(
                 std::make_tuple(std::vector<std::string>{"1"}, "NULL"),
                 std::make_tuple(std::vector<std::string>{R"("a")", "1", "1"}, R"(NULL)"),
                 std::make_tuple(std::vector<std::string>{R"("")"}, R"(NULL)")));
+
+TEST_F(JsonFunctionsTest, extract_from_object_test) {
+    std::string output;
+    Status st;
+
+    st = test_extract_from_object(R"({"data" : 1})", "$.data", &output);
+    EXPECT_OK(st);
+    EXPECT_STREQ(output.data(), "1");
+
+    st = test_extract_from_object(R"({"data" : 1})", "$.dataa", &output);
+    EXPECT_STATUS(Status::NotFound(""), st);
+
+    st = test_extract_from_object(R"({"data": [{"key": 1},{"key": 2}]})", "$.data[1].key", &output);
+    EXPECT_OK(st);
+    EXPECT_STREQ(output.data(), "2");
+
+    st = test_extract_from_object(R"({"data": [{"key": 1},{"key": 2}]})", "$.data[2].key", &output);
+    EXPECT_STATUS(Status::InvalidArgument(""), st);
+}
 
 } // namespace vectorized
 } // namespace starrocks

--- a/be/test/exprs/vectorized/json_functions_test.cpp
+++ b/be/test/exprs/vectorized/json_functions_test.cpp
@@ -48,7 +48,7 @@ public:
         JsonFunctions::parse_json_paths(jsonpath, &path);
 
         simdjson::ondemand::value val;
-        RETURN_IF_ERROR(JsonFunctions::extract_from_object(obj, path, val));
+        RETURN_IF_ERROR(JsonFunctions::extract_from_object(obj, path, &val));
         std::string_view sv = simdjson::to_json_string(val);
 
         output->assign(sv.data(), sv.size());

